website/src/config.md
---
title: Configuration
description: Documentation on dprint's configuration file.
layout: layouts/documentation.njk
---

# Configuration

Configuration allows you to modify how dprint and its plugins behave.

## Setup

See [Setup](/setup).

## Example

```json
{
  "lineWidth": 80,
  "typescript": {
    // This applies to both JavaScript & TypeScript
    "quoteStyle": "preferSingle",
    "binaryExpression.operatorPosition": "sameLine"
  },
  "json": {
    "indentWidth": 2
  },
  "excludes": [
    "**/*-lock.json"
  ],
  "plugins": [
    // You may specify any urls or file paths here that you wish.
    "https://plugins.dprint.dev/typescript-x.x.x.wasm",
    "https://plugins.dprint.dev/json-x.x.x.wasm",
    "https://plugins.dprint.dev/markdown-x.x.x.wasm"
  ]
}
```

## Plugins

The `plugins` property specifies which plugins to use for formatting. These may be URLs or file paths to a WebAssembly file of the plugin.

```json
{
  // ...omitted...
  "plugins": [
    // You may specify any urls or file paths here that you wish.
    "https://plugins.dprint.dev/typescript-x.x.x.wasm",
    "https://plugins.dprint.dev/json-x.x.x.wasm",
    "https://plugins.dprint.dev/markdown-x.x.x.wasm"
  ]
}
```

Alternatively, these may be provided to the CLI via the `--plugins <plugin urls or file paths...>` flag.

Note: The order of the plugins in this array defines the precedence. If two plugins support the same file extension then define the one you want to format that extension with first. For more fine grained control, see "associations" below.

### Adding Plugins via CLI

You may add a plugin via the CLI by running:

```sh
dprint config add
```

This will prompt asking you which of the standard plugins you want to add.

Alternatively, specify the specific name of the plugin to add based on its GitHub repo:

```sh
dprint config add dprint/dprint-plugin-typescript
```

Or for the standard plugins, you can just do:

```sh
dprint config add typescript
```

Or specify a plugin url:

```sh
dprint config add https://plugins.dprint.dev/json-x.x.x.wasm
```

### Updating Plugins via CLI

Plugins can be updated to the latest version in the configuration file by running:

```sh
dprint config update
```

## Excludes

The `excludes` property specifies the file paths exclude from formatting.

These should be file globs according to [`gitignore`'s extended glob syntax](https://git-scm.com/docs/gitignore#_pattern_format):

```json
{
  // ...omitted...
  "excludes": [
    "**/*-lock.json"
  ]
}
```

### Un-excluding gitignored files

Files that are gitignored will be excluded by default, but you can "un-exclude" them by specifying a negated glob:

```json
{
  "excludes": [
    // will format dist.js even though it's gitignored
    "!dist.js"
  ]
}
```

## Includes

The `includes` property can be used to limit dprint to only formatting certain files. Generally, you don't need to bother providing this.

```json
{
  // ...omitted...
  "includes": [
    "src/**/*.{ts,tsx,js,jsx,json}"
  ]
}
```

## Associations

By default, plugins will pull in files based on their extension. Sometimes a file may have a different extension or no extension at all, but you still want to format it with a certain plugin. The plugin `"associations"` config allows you to do that by associating a certain file pattern to one or multiple plugins.

For example:

```json
{
  "json": {
    "associations": [
      // format any file named `.myconfigrc` matched by the
      // includes/excludes patterns in any directory with
      // the json plugin
      ".myconfigrc",
      // format this specific file using the json plugin
      "./my-relative-path/to-file",
      // format files that match this pattern
      "**/*.myconfig"
    ]
  },
  "plugins": [
    "https://plugins.dprint.dev/json-x.x.x.wasm"
  ]
}
```

Note that first the `"includes"`/`"excludes"` file resolution occurs and then the associations is used to map those files to a plugin. Specifying associations may also be useful for formatting a file with multiple plugins or forcing a file to be formatted with a specific plugin.

### Excluding paths from plugin

Only providing negated globs as an association can be a way to exclude a file extension or path from being formatted with a certain plugin, but continue using file extensions to match a plugin otherwise.

In the following example, both the TypeScript plugin and Prettier plugin support formatting `.js` and `.ts` files. Say we want to only format `.ts` files with the TypeScript plugin and `.js` files with the prettier plugin. To do that, we can place the typescript plugin to have higher precedence in the "plugins" array, then add an excludes for only `!**/*.js`. This will cause the TypeScript plugin to match based on the file extension for `.ts` files, but then be excluded from matching on `.js` files.

```json
{
  "typescript": {
    "associations": [
      "!**/*.js" // don't format javascript files
    ]
  },
  "plugins": [
    "https://plugins.dprint.dev/typescript-x.x.x.wasm",
    // side note: check the docs for the latest version of this plugin
    "https://plugins.dprint.dev/prettier-0.13.0.json@dc5d12b7c1bf1a4683eff317c2c87350e75a5a3dfcc127f3d5628931bfb534b1"
  ]
}
```

## Extending a Different Configuration File

You may extend other configuration files by specifying an `extends` property. This may be a file path, URL, or relative path (remote configuration may extend other configuration files via a relative path).

<!-- dprint-ignore -->

```json
{
  "extends": "https://dprint.dev/path/to/config/file.v1.json",
  // ...omitted...
}
```

Referencing multiple configuration files is also supported. These should be ordered by precedence:

```json
{
  "extends": [
    "https://dprint.dev/path/to/config/file.v1.json",
    "https://dprint.dev/path/to/config/other.v1.json"
  ]
}
```

Note: The `includes` property of extended _remote_ configuration is ignored for security reasons out of an abundance of caution (to disallow the dprint cli pulling in sensitive files) and additionally non-Wasm plugins are ignored in remote configuration because they don't run sandboxed.

## Incremental

By default, dprint will only format files that have changed since the last time you formatted the code in order to drastically improve performance.

If you want to disable this functionality, you may specify the following in your dprint configuration file:

```json
{
  // etc...
  "incremental": false
  // etc...
}
```

Alternatively, specify `--incremental=false` on the CLI:

```sh
dprint fmt --incremental=false
```

## Common Configuration

There are certain non-language specific configuration that can be specified. These are specified on the main configuration object, but can be overridden on a per-plugin basis.

For example:

```json
{
  "lineWidth": 160,
  "useTabs": true,
  "typescript": {
    "lineWidth": 80
  },
  "json": {
    "indentWidth": 2,
    "useTabs": false
  },
  "plugins": [
    // etc...
  ]
}
```

Note: dprint does not provide or enforce any defaults for the common configuration. The defaults are set on a per plugin basis. When a value is not provided, the plugin may choose to use its default.

### `lineWidth`

The width of a line the formatter will try to stay under. Note that this limit will be exceeded in certain cases.

### `indentWidth`

The number of spaces for an indent when using spaces or the number of characters to treat an indent as when using tabs.

### `newLineKind`

The kind of newline to use.

- `auto` - For each file, uses the newline kind found at the end of the last line.
- `crlf` - Uses carriage return, line feed.
- `lf` - Uses line feed.
- `system` - Uses the system standard (ex. crlf on Windows).

### `useTabs`

Whether to use tabs (`true`) or spaces (`false`).

## Locking Configuration—Opinionated Configurations

You may want to publish your own opinionated configuration and disallow anyone using it from overriding the properties.

This can be done by adding a `"locked": true` property to each plugin configuration you wish to lock.

Note: When doing this, ensure you set all the global configuration values if you wish to enforce those.

### Example

Say the following configuration were published at `https://dprint.dev/configs/my-config.json`:

```json
{
  "typescript": {
    "locked": true,
    "lineWidth": 80,
    "indentWidth": 2,
    "useTabs": false,
    "quoteStyle": "preferSingle",
    "binaryExpression.operatorPosition": "sameLine"
  },
  "json": {
    "locked": true,
    "lineWidth": 80,
    "indentWidth": 2,
    "useTabs": false
  }
}
```

The following would work fine:

```json
{
  "extends": "https://dprint.dev/configs/my-config.json",
  "myOtherPlugin": {
    "propertySeparator": "comma"
  },
  "plugins": [
    "https://plugins.dprint.dev/typescript-x.x.x.wasm",
    "https://plugins.dprint.dev/json-x.x.x.wasm",
    "https://plugins.dprint.dev/my-other-plugin-0.1.0.wasm"
  ]
}
```

But specifying properties in the `"typescript"` or `"json"` objects would cause an error when running in the CLI:

```json
{
  "extends": "https://dprint.dev/configs/my-config.json",
  "typescript": {
    "useBraces": "always" // error, "typescript" config was locked
  },
  "json": {
    "lineWidth": 120 // error, "json" config was locked
  },
  "myOtherPlugin": {
    "propertySeparator": "comma"
  },
  "plugins": [
    "https://plugins.dprint.dev/typescript-x.x.x.wasm",
    "https://plugins.dprint.dev/json-x.x.x.wasm",
    "https://plugins.dprint.dev/my-other-plugin-0.1.0.wasm"
  ]
}
```

## Configuration Variables

Requires dprint >= 0.47.0

dprint expands certain variables in the config:

- `${configDir}` - The current configuration's directory.
- `${originConfigDir}` - The original configuration's directory. Useful when the current config is being extended by another configuration file and you want the original directory.

For example, in a JSON value you might do `"rustfmt --config-path ${configDir}/rustfmt.toml"`.

This is useful to use in some scenarios like with [dprint-plugin-exec](https://github.com/dprint/dprint-plugin-exec) because the CLI will only launch a single plugin for many configs and when resolving configs, the plugins have no concept of where that config was resolved from. Additionally, configs may resolve other configs and perhaps you want to use the directory of a configuration file that was extended.

Note: dprint will error for unknown configuration variables (ex. `"${unknown}"`). You can get around this by escaping the `$` sign (ex. `"\\${unknown}"`).

## Plugin/Language Specific Configuration

Running `dprint help` will list the help urls for all the configured plugins in your configuration file. On those pages you can view the help information.

For information on the official plugins' configuration, see the [plugins](https://dprint.dev/plugins/) section.

Next step: [CLI](/cli)


crates/core/src/configuration.rs
use std::hash::Hash;

use indexmap::IndexMap;
use serde::Deserialize;
use serde::Serialize;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParseConfigurationError(pub String);

impl std::fmt::Display for ParseConfigurationError {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    format!("Found invalid value '{}'.", self.0).fmt(f)
  }
}

#[macro_export]
macro_rules! generate_str_to_from {
    ($enum_name:ident, $([$member_name:ident, $string_value:expr]),* ) => {
        impl std::str::FromStr for $enum_name {
            type Err = ParseConfigurationError;

            fn from_str(s: &str) -> Result<Self, Self::Err> {
                match s {
                    $($string_value => Ok($enum_name::$member_name)),*,
                    _ => Err(ParseConfigurationError(String::from(s))),
                }
            }
        }

        impl std::fmt::Display for $enum_name {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                match self {
                    $($enum_name::$member_name => write!(f, $string_value)),*,
                }
            }
        }
    };
}

#[derive(Clone, PartialEq, Eq, Debug, Copy, Serialize, Deserialize, Hash)]
pub enum RawNewLineKind {
  /// Decide which newline kind to use based on the last newline in the file.
  #[serde(rename = "auto")]
  Auto,
  /// Use slash n new lines.
  #[serde(rename = "lf")]
  LineFeed,
  /// Use slash r slash n new lines.
  #[serde(rename = "crlf")]
  CarriageReturnLineFeed,
  /// Use the system standard (ex. crlf on Windows)
  #[serde(rename = "system")]
  System,
}

generate_str_to_from![
  RawNewLineKind,
  [Auto, "auto"],
  [LineFeed, "lf"],
  [CarriageReturnLineFeed, "crlf"],
  [System, "system"]
];

#[derive(Clone, PartialEq, Eq, Debug, Copy, Serialize, Deserialize, Hash)]
pub enum NewLineKind {
  /// Decide which newline kind to use based on the last newline in the file.
  #[serde(rename = "auto")]
  Auto,
  /// Use slash n new lines.
  #[serde(rename = "lf")]
  LineFeed,
  /// Use slash r slash n new lines.
  #[serde(rename = "crlf")]
  CarriageReturnLineFeed,
}

generate_str_to_from![NewLineKind, [Auto, "auto"], [LineFeed, "lf"], [CarriageReturnLineFeed, "crlf"]];

/// Represents a problem within the configuration.
#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct ConfigurationDiagnostic {
  /// The property name the problem occurred on.
  pub property_name: String,
  /// The diagnostic message that should be displayed to the user
  pub message: String,
}

impl std::fmt::Display for ConfigurationDiagnostic {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    write!(f, "{} ({})", self.message, self.property_name)
  }
}

pub type ConfigKeyMap = IndexMap<String, ConfigKeyValue>;

#[derive(Clone, PartialEq, Eq, Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ConfigKeyValue {
  String(String),
  Number(i32),
  Bool(bool),
  Array(Vec<ConfigKeyValue>),
  Object(ConfigKeyMap),
  Null,
}

impl ConfigKeyValue {
  pub fn as_string(&self) -> Option<&String> {
    match self {
      ConfigKeyValue::String(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_number(&self) -> Option<i32> {
    match self {
      ConfigKeyValue::Number(value) => Some(*value),
      _ => None,
    }
  }

  pub fn as_bool(&self) -> Option<bool> {
    match self {
      ConfigKeyValue::Bool(value) => Some(*value),
      _ => None,
    }
  }

  pub fn as_array(&self) -> Option<&Vec<ConfigKeyValue>> {
    match self {
      ConfigKeyValue::Array(values) => Some(values),
      _ => None,
    }
  }

  pub fn as_object(&self) -> Option<&ConfigKeyMap> {
    match self {
      ConfigKeyValue::Object(values) => Some(values),
      _ => None,
    }
  }

  pub fn into_string(self) -> Option<String> {
    match self {
      ConfigKeyValue::String(value) => Some(value),
      _ => None,
    }
  }

  pub fn into_number(self) -> Option<i32> {
    match self {
      ConfigKeyValue::Number(value) => Some(value),
      _ => None,
    }
  }

  pub fn into_bool(self) -> Option<bool> {
    match self {
      ConfigKeyValue::Bool(value) => Some(value),
      _ => None,
    }
  }

  pub fn into_array(self) -> Option<Vec<ConfigKeyValue>> {
    match self {
      ConfigKeyValue::Array(values) => Some(values),
      _ => None,
    }
  }

  pub fn into_object(self) -> Option<ConfigKeyMap> {
    match self {
      ConfigKeyValue::Object(values) => Some(values),
      _ => None,
    }
  }

  pub fn is_null(&self) -> bool {
    matches!(self, ConfigKeyValue::Null)
  }

  /// Gets a hash of the configuration value. This is used for incremental formatting
  /// and the Hash trait is not implemented to discourage using this in other places.
  #[allow(clippy::should_implement_trait)]
  pub fn hash(&self, hasher: &mut impl std::hash::Hasher) {
    match self {
      ConfigKeyValue::String(value) => {
        hasher.write_u8(0);
        hasher.write(value.as_bytes())
      }
      ConfigKeyValue::Number(value) => {
        hasher.write_u8(1);
        hasher.write_i32(*value)
      }
      ConfigKeyValue::Bool(value) => {
        hasher.write_u8(2);
        hasher.write_u8(if *value { 1 } else { 0 })
      }
      ConfigKeyValue::Array(values) => {
        hasher.write_u8(3);
        for value in values {
          value.hash(hasher);
        }
      }
      ConfigKeyValue::Object(key_values) => {
        hasher.write_u8(4);
        for (key, value) in key_values {
          hasher.write(key.as_bytes());
          value.hash(hasher);
        }
      }
      ConfigKeyValue::Null => {
        hasher.write_u8(5);
      }
    }
  }

  pub fn from_i32(value: i32) -> ConfigKeyValue {
    ConfigKeyValue::Number(value)
  }

  #[allow(clippy::should_implement_trait)]
  pub fn from_str(value: &str) -> ConfigKeyValue {
    ConfigKeyValue::String(value.to_string())
  }

  pub fn from_bool(value: bool) -> ConfigKeyValue {
    ConfigKeyValue::Bool(value)
  }
}

impl From<i32> for ConfigKeyValue {
  fn from(item: i32) -> Self {
    ConfigKeyValue::from_i32(item)
  }
}

impl From<bool> for ConfigKeyValue {
  fn from(item: bool) -> Self {
    ConfigKeyValue::from_bool(item)
  }
}

impl From<String> for ConfigKeyValue {
  fn from(item: String) -> Self {
    ConfigKeyValue::from_str(&item)
  }
}

impl From<&str> for ConfigKeyValue {
  fn from(item: &str) -> Self {
    ConfigKeyValue::from_str(item)
  }
}

#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, Debug, Default, Hash)]
#[serde(rename_all = "camelCase")]
pub struct CommonConfiguration {
  pub line_width: Option<u32>,
  pub use_tabs: Option<bool>,
  pub indent_width: Option<u8>,
  pub new_line_kind: Option<NewLineKind>,
}

pub const RECOMMENDED_COMMON_CONFIGURATION: RecommendedCommonConfiguration = RecommendedCommonConfiguration {
  line_width: 120,
  indent_width: 2,
  use_tabs: false,
  new_line_kind: NewLineKind::LineFeed,
};

pub struct RecommendedCommonConfiguration {
  pub line_width: u32,
  pub use_tabs: bool,
  pub indent_width: u8,
  pub new_line_kind: NewLineKind,
}

impl From<RecommendedCommonConfiguration> for CommonConfiguration {
  fn from(config: RecommendedCommonConfiguration) -> Self {
    Self {
      line_width: Some(config.line_width),
      use_tabs: Some(config.use_tabs),
      indent_width: Some(config.indent_width),
      new_line_kind: Some(config.new_line_kind),
    }
  }
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResolveConfigurationResult<T>
where
  T: Clone + Serialize,
{
  /// The configuration diagnostics.
  pub diagnostics: Vec<ConfigurationDiagnostic>,

  /// The configuration derived from the unresolved configuration
  /// that can be used to format a file.
  pub config: T,
}

/// Resolves a collection of key value pairs to a CommonConfiguration.
pub fn resolve_common_config(config: &mut ConfigKeyMap) -> ResolveConfigurationResult<CommonConfiguration> {
  let mut diagnostics = Vec::new();

  let raw_new_line_kind = get_nullable_value::<RawNewLineKind>(config, "newLineKind", &mut diagnostics);

  let resolved_config = CommonConfiguration {
    line_width: get_nullable_value(config, "lineWidth", &mut diagnostics),
    use_tabs: get_nullable_value(config, "useTabs", &mut diagnostics),
    indent_width: get_nullable_value(config, "indentWidth", &mut diagnostics),
    new_line_kind: raw_new_line_kind.map(|kind| match kind {
      RawNewLineKind::Auto => NewLineKind::Auto,
      RawNewLineKind::LineFeed => NewLineKind::LineFeed,
      RawNewLineKind::CarriageReturnLineFeed => NewLineKind::CarriageReturnLineFeed,
      RawNewLineKind::System => {
        if cfg!(windows) {
          NewLineKind::CarriageReturnLineFeed
        } else {
          NewLineKind::LineFeed
        }
      }
    }),
  };

  ResolveConfigurationResult {
    config: resolved_config,
    diagnostics,
  }
}

/// If the provided key exists, takes its value from the provided config and returns it.
/// If the provided key does not exist, it returns the default value.
/// Adds a diagnostic if there is any problem deserializing the value.
pub fn get_value<T>(config: &mut ConfigKeyMap, key: &str, default_value: T, diagnostics: &mut Vec<ConfigurationDiagnostic>) -> T
where
  T: std::str::FromStr,
  <T as std::str::FromStr>::Err: std::fmt::Display,
{
  get_nullable_value(config, key, diagnostics).unwrap_or(default_value)
}

/// If the provided key exists, takes its value from the provided config and returns it.
/// If the provided key does not exist, it returns None.
/// Adds a diagnostic if there is any problem deserializing the value.
pub fn get_nullable_value<T>(config: &mut ConfigKeyMap, key: &str, diagnostics: &mut Vec<ConfigurationDiagnostic>) -> Option<T>
where
  T: std::str::FromStr,
  <T as std::str::FromStr>::Err: std::fmt::Display,
{
  if let Some(raw_value) = config.shift_remove(key) {
    // not exactly the best, but can't think of anything better at the moment
    let parsed_value = match raw_value {
      ConfigKeyValue::Bool(value) => value.to_string().parse::<T>().map_err(|e| e.to_string()),
      ConfigKeyValue::Number(value) => value.to_string().parse::<T>().map_err(|e| e.to_string()),
      ConfigKeyValue::String(value) => value.parse::<T>().map_err(|e| e.to_string()),
      ConfigKeyValue::Object(_) | ConfigKeyValue::Array(_) => Err("Arrays and objects are not supported for this value".to_string()),
      ConfigKeyValue::Null => return None,
    };
    match parsed_value {
      Ok(parsed_value) => Some(parsed_value),
      Err(message) => {
        diagnostics.push(ConfigurationDiagnostic {
          property_name: key.to_string(),
          message,
        });
        None
      }
    }
  } else {
    None
  }
}

pub fn get_nullable_vec<T: std::str::FromStr>(
  config: &mut ConfigKeyMap,
  key: &str,
  get_nullable_value: impl Fn(ConfigKeyValue, usize, &mut Vec<ConfigurationDiagnostic>) -> Option<T>,
  diagnostics: &mut Vec<ConfigurationDiagnostic>,
) -> Option<Vec<T>> {
  match config.shift_remove(key) {
    Some(value) => match value {
      ConfigKeyValue::Array(values) => {
        let mut result = Vec::with_capacity(values.len());
        for (i, value) in values.into_iter().enumerate() {
          if let Some(value) = get_nullable_value(value, i, diagnostics) {
            result.push(value);
          }
        }
        Some(result)
      }
      _ => {
        diagnostics.push(ConfigurationDiagnostic {
          property_name: key.to_string(),
          message: "Expected an array.".to_string(),
        });
        None
      }
    },
    None => None,
  }
}

/// If it exists, moves over the configuration value over from the old key
/// to the new key and adds a diagnostic.
pub fn handle_renamed_config_property(config: &mut ConfigKeyMap, old_key: &str, new_key: &str, diagnostics: &mut Vec<ConfigurationDiagnostic>) {
  if let Some(raw_value) = config.shift_remove(old_key) {
    if !config.contains_key(new_key) {
      config.insert(new_key.to_string(), raw_value);
    }
    diagnostics.push(ConfigurationDiagnostic {
      property_name: old_key.to_string(),
      message: format!("The configuration key was renamed to '{}'", new_key),
    });
  }
}

/// Resolves the `NewLineKind` text from the provided file text and `NewLineKind`.
pub fn resolve_new_line_kind(file_text: &str, new_line_kind: NewLineKind) -> &'static str {
  match new_line_kind {
    NewLineKind::LineFeed => "\n",
    NewLineKind::CarriageReturnLineFeed => "\r\n",
    NewLineKind::Auto => {
      let mut found_slash_n = false;
      for c in file_text.as_bytes().iter().rev() {
        if found_slash_n {
          if c == &(b'\r') {
            return "\r\n";
          } else {
            return "\n";
          }
        }

        if c == &(b'\n') {
          found_slash_n = true;
        }
      }

      "\n"
    }
  }
}

/// Gets a diagnostic for each remaining key value pair in the hash map.
///
/// This should be done last, so it swallows the hashmap.
pub fn get_unknown_property_diagnostics(config: ConfigKeyMap) -> Vec<ConfigurationDiagnostic> {
  let mut diagnostics = Vec::new();
  for (key, _) in config {
    diagnostics.push(ConfigurationDiagnostic {
      property_name: key.to_string(),
      message: "Unknown property in configuration".to_string(),
    });
  }
  diagnostics
}

#[cfg(test)]
mod test {
  use super::*;

  #[test]
  fn get_default_config_when_empty() {
    let config_result = resolve_common_config(&mut ConfigKeyMap::new());
    let config = config_result.config;
    assert_eq!(config_result.diagnostics.len(), 0);
    assert_eq!(config.line_width, None);
    assert_eq!(config.indent_width, None);
    assert!(config.new_line_kind.is_none());
    assert_eq!(config.use_tabs, None);
  }

  #[test]
  fn get_values_when_filled() {
    let mut common_config = ConfigKeyMap::from([
      (String::from("lineWidth"), ConfigKeyValue::from_i32(80)),
      (String::from("indentWidth"), ConfigKeyValue::from_i32(8)),
      (String::from("newLineKind"), ConfigKeyValue::from_str("crlf")),
      (String::from("useTabs"), ConfigKeyValue::from_bool(true)),
    ]);
    let config_result = resolve_common_config(&mut common_config);
    let config = config_result.config;
    assert_eq!(config_result.diagnostics.len(), 0);
    assert_eq!(config.line_width, Some(80));
    assert_eq!(config.indent_width, Some(8));
    assert_eq!(config.new_line_kind, Some(NewLineKind::CarriageReturnLineFeed));
    assert_eq!(config.use_tabs, Some(true));
  }

  #[test]
  fn get_diagnostic_for_invalid_enum_config() {
    let mut common_config = ConfigKeyMap::from([(String::from("newLineKind"), ConfigKeyValue::from_str("something"))]);
    let diagnostics = resolve_common_config(&mut common_config).diagnostics;
    assert_eq!(diagnostics.len(), 1);
    assert_eq!(diagnostics[0].message, "Found invalid value 'something'.");
    assert_eq!(diagnostics[0].property_name, "newLineKind");
  }

  #[test]
  fn get_diagnostic_for_invalid_primitive() {
    let mut common_config = ConfigKeyMap::from([(String::from("useTabs"), ConfigKeyValue::from_str("something"))]);
    let diagnostics = resolve_common_config(&mut common_config).diagnostics;
    assert_eq!(diagnostics.len(), 1);
    assert_eq!(diagnostics[0].message, "provided string was not `true` or `false`");
    assert_eq!(diagnostics[0].property_name, "useTabs");
  }

  #[test]
  fn get_diagnostic_for_excess_property() {
    let common_config = ConfigKeyMap::from([(String::from("something"), ConfigKeyValue::from_str("value"))]);
    let diagnostics = get_unknown_property_diagnostics(common_config);
    assert_eq!(diagnostics.len(), 1);
    assert_eq!(diagnostics[0].message, "Unknown property in configuration");
    assert_eq!(diagnostics[0].property_name, "something");
  }

  #[test]
  fn add_diagnostic_for_renamed_property() {
    let mut config = ConfigKeyMap::new();
    let mut diagnostics = Vec::new();
    config.insert("oldProp".to_string(), ConfigKeyValue::from_str("value"));
    handle_renamed_config_property(&mut config, "oldProp", "newProp", &mut diagnostics);
    assert_eq!(config.len(), 1);
    assert_eq!(config.shift_remove("newProp").unwrap(), ConfigKeyValue::from_str("value"));
    assert_eq!(diagnostics.len(), 1);
    assert_eq!(diagnostics[0].message, "The configuration key was renamed to 'newProp'");
    assert_eq!(diagnostics[0].property_name, "oldProp");
  }

  #[test]
  fn add_diagnostic_for_renamed_property_when_already_exists() {
    let mut config = ConfigKeyMap::new();
    let mut diagnostics = Vec::new();
    config.insert("oldProp".to_string(), ConfigKeyValue::from_str("new_value"));
    config.insert("newProp".to_string(), ConfigKeyValue::from_str("value"));
    handle_renamed_config_property(&mut config, "oldProp", "newProp", &mut diagnostics);
    assert_eq!(config.len(), 1);
    assert_eq!(config.shift_remove("newProp").unwrap(), ConfigKeyValue::from_str("value"));
    assert_eq!(diagnostics.len(), 1);
    assert_eq!(diagnostics[0].message, "The configuration key was renamed to 'newProp'");
    assert_eq!(diagnostics[0].property_name, "oldProp");
  }
}
